From 1cc62b8fa4947e48183624441465d7421ab57708 Mon Sep 17 00:00:00 2001
From: Manas <manas@Manass-MacBook-Pro.local>
Date: Mon, 28 Nov 2022 13:32:47 +0530
Subject: [PATCH] Removes problematic sh script on macos

---
 dune | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dune b/dune
index fc85f4a40..eefcd5315 100644
--- a/dune
+++ b/dune
@@ -12,4 +12,4 @@
 (rule
  (target static-link-flags.sexp)
  (action (with-stdout-to %{target}
-  (system "[ '%{profile}' = 'static' ] && echo '(-ccopt -static -cclib -lusb-1.0 -cclib -ludev)' || echo ()"))))
+  (system "echo '()'"))))
-- 
2.34.1

