From a7752fff7c0531e767e339c9e8154029ae3d6a0a Mon Sep 17 00:00:00 2001
From: Manas Jayanth <3097018+ManasJayanth@users.noreply.github.com>
Date: Thu, 13 Jan 2022 07:58:48 +0530
Subject: [PATCH] Adds win32 guards

---
 src/rusage_stubs.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/rusage_stubs.c b/src/rusage_stubs.c
index 9839970..14b5e81 100644
--- a/src/rusage_stubs.c
+++ b/src/rusage_stubs.c
@@ -19,12 +19,15 @@ PERFORMANCE OF THIS SOFTWARE. */
 #include <caml/memory.h>
 #include <caml/mlvalues.h>
 
+#if ! (defined(WIN32) || defined(_WIN32))
 #include <sys/resource.h>
 #include <sys/time.h>
+#endif
 
 #define Nothing Val_int(0)
 
 CAMLprim value unix_getrusage(value v_who) {
+#if ! (defined(WIN32) || defined(_WIN32))
   CAMLparam1(v_who);
   CAMLlocal1(v_usage);
   int who = (Int_val(v_who) == 0) ? RUSAGE_SELF : RUSAGE_CHILDREN;
@@ -54,4 +57,8 @@ CAMLprim value unix_getrusage(value v_who) {
   Store_field(v_usage, 14, caml_copy_int64(ru.ru_nvcsw));
   Store_field(v_usage, 15, caml_copy_int64(ru.ru_nivcsw));
   CAMLreturn(v_usage);
+#else
+  caml_failwith("not implemented for win32");
+  return Val_unit;
+#endif
 }
-- 
2.34.1.windows.1

