--- ./src/dune
+++ ./src/dune
@@ -8,6 +8,7 @@
 
 (library
  (name class_group_vdf)
+ (no_dynlink)
  (public_name class_group_vdf)
  (libraries integers zarith)
  (modules Class_group Vdf Utils S Vdf_self_contained)
--- ./src/libvdf/src/alloc.hpp
+++ ./src/libvdf/src/alloc.hpp
@@ -3,6 +3,34 @@
 
 #include <stdlib.h> // for posix_memalign
 
+#if (defined(WIN32) || defined(_WIN32))
+static int check_align(size_t align)
+{
+    for (size_t i = sizeof(void *); i != 0; i *= 2)
+    if (align == i)
+        return 0;
+    return EINVAL;
+}
+
+int posix_memalign(void **ptr, size_t align, size_t size)
+{
+    if (check_align(align))
+        return EINVAL;
+
+    int saved_errno = errno;
+    void *p = _aligned_malloc(size, align);
+    if (p == NULL)
+    {
+        errno = saved_errno;
+        return ENOMEM;
+    }
+
+    *ptr = p;
+    return 0;
+}
+#endif
+
+
 inline void* mp_alloc_func(size_t new_bytes)
 {
     new_bytes = ((new_bytes + 8) + 15) & ~15;

